File: server/ClipboardServer.java
import java.awt.Toolkit;
import java.awt.datatransfer.Clipboard;
import java.awt.datatransfer.StringSelection;
import java.io.BufferedReader;
import java.net.DatagramPacket;
import java.net.DatagramSocket;
import java.net.Inet4Address;
import java.net.InetAddress;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.nio.charset.StandardCharsets;
import java.net.NetworkInterface;
import java.net.ServerSocket;
import java.net.Socket;
import java.util.Enumeration;
import java.util.Locale;

public class ClipboardServer {
    private static final int PORT = 8888;
    private static final int DISCOVERY_PORT = 9999;
    private static final String MAC_NAME = "mac";
    private static final String DISCOVER_MESSAGE = "SMSSYNC_DISCOVER";
    private static final String SERVER_RESPONSE_PREFIX = "SMSSYNC_SERVER";

    public static void main(String[] args) {
        startDiscoveryServer();

        try (ServerSocket serverSocket = new ServerSocket(PORT)) {
            serverSocket.setReuseAddress(true);
            System.out.println("服务端启动！正在监听端口 " + PORT);
            System.out.println("本机 IP 地址 ");
            printRealIp();
            
            while (true) {
                handleClient(serverSocket);
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    private static void startDiscoveryServer() {
        Thread discoveryThread = new Thread(new Runnable() {
            @Override
            public void run() {
                try (DatagramSocket discoverySocket = new DatagramSocket(DISCOVERY_PORT)) {
                    discoverySocket.setBroadcast(true);
                    System.out.println("自动发现服务启动！正在监听端口 " + DISCOVERY_PORT);

                    byte[] buffer = new byte[256];
                    while (true) {
                        DatagramPacket packet = new DatagramPacket(buffer, buffer.length);
                        discoverySocket.receive(packet);
                        handleDiscoveryRequest(discoverySocket, packet);
                    }
                } catch (Exception e) {
                    e.printStackTrace();
                }
            }
        });

        discoveryThread.setDaemon(true);
        discoveryThread.start();
    }

    private static void handleDiscoveryRequest(DatagramSocket discoverySocket, DatagramPacket packet) {
        try {
            String message = new String(packet.getData(), 0, packet.getLength(), StandardCharsets.UTF_8).trim();
            if (!DISCOVER_MESSAGE.equals(message)) {
                return;
            }

            String serverIp = getLocalIpForClient(packet.getAddress());
            if (serverIp == null || serverIp.isEmpty()) {
                return;
            }

            String response = SERVER_RESPONSE_PREFIX + "|" + serverIp + "|" + PORT;
            byte[] responseData = response.getBytes(StandardCharsets.UTF_8);
            DatagramPacket responsePacket = new DatagramPacket(
                    responseData,
                    responseData.length,
                    packet.getAddress(),
                    packet.getPort()
            );

            discoverySocket.send(responsePacket);
            System.out.println("已回复自动发现请求: " + serverIp);
        } catch (Exception e) {
            System.err.println(e.getMessage());
        }
    }

    private static void handleClient(ServerSocket serverSocket) {
        try (Socket clientSocket = serverSocket.accept();
             BufferedReader reader = new BufferedReader(
                     new InputStreamReader(clientSocket.getInputStream(), StandardCharsets.UTF_8))) {
            String message = reader.readLine();
            if (message == null || message.isEmpty()) {
                return;
            }

            System.out.println("接收到消息: " + message);
            setClipboard(message);
        } catch (Exception e) {
            System.err.println(e.getMessage());
        }
    }

    private static String getLocalIpForClient(InetAddress clientAddress) {
        try (DatagramSocket socket = new DatagramSocket()) {
            socket.connect(clientAddress, DISCOVERY_PORT);
            InetAddress localAddress = socket.getLocalAddress();
            if (localAddress instanceof Inet4Address && !localAddress.isLoopbackAddress()) {
                return localAddress.getHostAddress();
            }
        } catch (Exception e) {
        }

        return getFirstIpv4Address();
    }

    private static void printRealIp() {
        String firstIp = getFirstIpv4Address();
        if (firstIp != null) {
            System.out.println("  " + firstIp);
        }

        try {
            Enumeration<NetworkInterface> interfaces = NetworkInterface.getNetworkInterfaces();
            while (interfaces.hasMoreElements()) {
                NetworkInterface iface = interfaces.nextElement();
                if (iface.isLoopback() || !iface.isUp()) {
                    continue;
                }

                Enumeration<InetAddress> addresses = iface.getInetAddresses();
                while (addresses.hasMoreElements()) {
                    InetAddress addr = addresses.nextElement();
                    if (addr instanceof Inet4Address) {
                        String ip = addr.getHostAddress();
                        if (!ip.equals(firstIp)) {
                            System.out.println("  " + ip);
                        }
                    }
                }
            }
        } catch (Exception e) {
        }
    }

    private static String getFirstIpv4Address() {
        try {
            Enumeration<NetworkInterface> interfaces = NetworkInterface.getNetworkInterfaces();
            while (interfaces.hasMoreElements()) {
                NetworkInterface iface = interfaces.nextElement();
                if (iface.isLoopback() || !iface.isUp()) {
                    continue;
                }

                Enumeration<InetAddress> addresses = iface.getInetAddresses();
                while (addresses.hasMoreElements()) {
                    InetAddress addr = addresses.nextElement();
                    if (addr instanceof Inet4Address) {
                        return addr.getHostAddress();
                    }
                }
            }
        } catch (Exception e) {
        }

        return null;
    }

    private static void setClipboard(String text) {
        if (isMac()) {
            setClipboardOnMac(text);
        } else {
            StringSelection selection = new StringSelection(text);
            Clipboard clipboard = Toolkit.getDefaultToolkit().getSystemClipboard();
            clipboard.setContents(selection, selection);
        }
        System.out.println("成功复制到剪贴板！");
    }

    private static boolean isMac() {
        String osName = System.getProperty("os.name", "");
        return osName.toLowerCase(Locale.ROOT).contains(MAC_NAME);
    }

    private static void setClipboardOnMac(String text) {
        try {
            Process process = new ProcessBuilder("pbcopy").start();
            try (OutputStream outputStream = process.getOutputStream()) {
                outputStream.write(text.getBytes(StandardCharsets.UTF_8));
            }

            int exitCode = process.waitFor();
            if (exitCode != 0) {
                throw new RuntimeException("pbcopy exited with code " + exitCode);
            }
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }
}

File: server/ClipboardServer.py
import socket
import threading
import subprocess
import sys

PORT = 8888
DISCOVERY_PORT = 9999
DISCOVER_MESSAGE = "SMSSYNC_DISCOVER"
SERVER_RESPONSE_PREFIX = "SMSSYNC_SERVER"

def get_local_ips():
    ips = []
    try:
        host_name = socket.gethostname()
        for ip in socket.gethostbyname_ex(host_name)[2]:
            if not ip.startswith("127."):
                ips.append(ip)
    except Exception:
        pass
    
    try:
        s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
        s.connect(("8.8.8.8", 80))
        ip = s.getsockname()[0]
        s.close()
        if ip not in ips and not ip.startswith("127."):
            ips.insert(0, ip)
    except Exception:
        pass
        
    return ips

def is_mac():
    return sys.platform == 'darwin'

def set_clipboard(text):
    if is_mac():
        try:
            p = subprocess.Popen(['pbcopy'], stdin=subprocess.PIPE)
            p.communicate(input=text.encode('utf-8'))
        except Exception as e:
            print(f"剪贴板错误: {e}")
    else:
        if sys.platform == 'win32':
            try:
                p = subprocess.Popen(['clip'], stdin=subprocess.PIPE)
                p.communicate(input=text.encode('utf-16le'))
            except:
                pass
        else:
            try:
                p = subprocess.Popen(['xclip', '-selection', 'clipboard'], stdin=subprocess.PIPE)
                p.communicate(input=text.encode('utf-8'))
            except:
                pass
    print("成功复制到剪贴板！")

def handle_client(conn, addr):
    with conn:
        data = conn.recv(1024)
        if data:
            message = data.decode('utf-8').strip()
            if message:
                print(f"接收到消息: {message}")
                set_clipboard(message)

def tcp_server():
    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
        s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        s.bind(('', PORT))
        s.listen()
        print(f"服务端启动！正在监听端口 {PORT}")
        print("本机 IP 地址:")
        for ip in get_local_ips():
            print(f"  {ip}")
            
        while True:
            conn, addr = s.accept()
            threading.Thread(target=handle_client, args=(conn, addr), daemon=True).start()

def discovery_server():
    with socket.socket(socket.AF_INET, socket.SOCK_DGRAM) as s:
        s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        s.bind(('', DISCOVERY_PORT))
        print(f"自动发现服务启动！正在监听端口 {DISCOVERY_PORT}")
        
        while True:
            data, addr = s.recvfrom(1024)
            message = data.decode('utf-8').strip()
            if message == DISCOVER_MESSAGE:
                ips = get_local_ips()
                server_ip = ips[0] if ips else "127.0.0.1"
                
                client_prefix = ".".join(addr[0].split('.')[:3])
                for ip in ips:
                    if ip.startswith(client_prefix):
                        server_ip = ip
                        break
                        
                response = f"{SERVER_RESPONSE_PREFIX}|{server_ip}|{PORT}"
                s.sendto(response.encode('utf-8'), addr)
                print(f"发现请求来自: {addr[0]}")
                print(f"已回复自动发现请求: {server_ip}")

if __name__ == '__main__':
    threading.Thread(target=discovery_server, daemon=True).start()
    tcp_server()
